### PR TITLE
webrtc: wait for FIN_ACK before closing data channels

### PR DESCRIPTION
**Pull request by sukunrt**
part of #2656 

This removes the control message queue because that only ever adds 500 bytes to the queue. This will not cause too much HOL Blocking for other streams. 

There is no problem with backpressure since sctp writes never actually block. There is no blocking on sctp Write.

To verify:
```
func TestStreamBackpressure(t *testing.T) {
	client, _ := getDetachedDataChannels(t)

	clientStr := newStream(client.dc, client.rwc, nil)

	b := make([]byte, 1024)
	for i := 0; i < 1e7; i++ {
		n, err := clientStr.dataChannel.Write(b)
		if n != 1024 {
			t.Fatalf("failed to write to datachannel")
		}
		if err != nil {
			t.Fatalf("failed to write to datachannel")
		}
	}
}

```

### CODE REVIEW

File: p2p/test/transport/transport_test.go
Comment (marten-seemann) [inline p2p/test/transport/transport_test.go:317]: Exciting!

File: p2p/transport/webrtc/stream.go
Comment (marten-seemann) [inline p2p/transport/webrtc/stream.go:89]: Is there anything we can change / simplify here after 290 was merged?

File: p2p/transport/webrtc/stream.go
Comment (marten-seemann) [inline p2p/transport/webrtc/stream.go:68]: Nit: Don't call it once, since that's usually used for a `sync.Once`. Alternative: `readerSem`, to make clear that it's a semaphore.

**Review by marten-seemann — COMMENTED** [review]
Partial review.

Discussed offline: If we move closing of the stream scope to the transports, we don't need to introduce the `AsyncCloser` interface. We need to decide if that structure makes sense first.

File: p2p/transport/webrtc/connection.go
Comment (sukunrt) [reply p2p/transport/webrtc/connection.go:140]: Good point. We don't need to, pion does handle this correctly. I'll remove this.

**Comment by sukunrt** [comment]
I don't have a good idea on how to ensure that we collect the stream handler goroutine and then close the stream rcmgr scope if the transport stream object is responsible for closing the scope and not the swarm stream object. 

However a public interface that exposes AsyncClose doesn't seem required either. I've made the assertion private to swarm package. This is subtle action at a distance but at least it makes it clear that we don't want this AsyncClose system to be copied in other transports.

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go:95]: This hasn't been merged yet. 


When it's done we can remove the `controlMsgReaderEndTime` and just set the deadline in AsyncClose & Close to be now + 10 seconds. We currently cannot do that because there's no way of knowing whether a deadlineexceeded error occurred because of a past canceled deadline or because of the most recently set deadline.

File: p2p/net/swarm/swarm_stream.go
Comment (Jorropo) [inline p2p/net/swarm/swarm_stream.go:79]: Add docs on the method please, I had to read the implementation to understand this actually also started the close process.

File: p2p/net/swarm/swarm_stream_test.go
Comment (Jorropo) [inline p2p/net/swarm/swarm_stream_test.go:21]: Could we throw a `go` statement somewhere to exercise potential deadlocks and races ?
I think it would be doable by adding `sync.WaitGroup` inside `asyncStreamWrapper`.

We could also add a second `sync.Mutex` which would allow the test the behavior of listing streams while an `AsyncClose` is still in flight (does that deadlocks ? Can we still find the open stream ?).

File: p2p/net/swarm/swarm_stream.go
Comment (Jorropo) [inline p2p/net/swarm/swarm_stream.go:90]: This means consumers can't acutally know if the remote ACKed received data (since `.Write` does not block for acks).
But I checked and we don't promise this in the first place:
> ```go
> // Close may be asynchronous and _does not_ guarantee receipt of the data.
> ```

I guess it's fine then.

File: p2p/transport/webrtc/stream.go
Comment (Jorropo) [inline p2p/transport/webrtc/stream.go:152]: Shouldn't this error if reading the ack fails ?
Like you can't actually fail to close a stream (unless maybe it's already closed) because even if the remote refuse to close you would still stop responding on this stream, so I always thought the `error` field was used to indicate already written data has been lost.

File: p2p/transport/webrtc/transport.go
Comment (Jorropo) [inline p2p/transport/webrtc/transport.go:80]: Pion defaults to 1460 which sounds correct:

https://github.com/pion/webrtc/blob/c259e89eb53b6d599c9dc7ec3e259fdb26e15f95/constants.go#L9-L10

Can you add a comment please (or remove it entirely if it's not needed) ?

**Review by Jorropo — COMMENTED** [review]
I havn't reviewed everything yet, I have a few questions.

File: p2p/transport/webrtc/transport.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport.go:80]: This failed once with ErrShortBuffer when the MTU was only 1460. But I think that's because this has a higher MTU and we should remove both the settings(https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/webrtc/udpmux/mux.go#L20)

File: p2p/net/swarm/swarm_stream.go
Comment (sukunrt) [reply p2p/net/swarm/swarm_stream.go:90]: yeah, there is no guarantee that the data will be delivered after Close so the semantics seem okay.

File: p2p/transport/webrtc/transport.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport.go]: I've removed this for now. I'll investigate this when this fails.

File: p2p/net/swarm/swarm_stream_test.go
Comment (sukunrt) [reply p2p/net/swarm/swarm_stream_test.go]: I updated it with test for 100 concurrent stream Close with GetStreams check. Is this what you had in mind?

File: p2p/net/swarm/swarm_stream.go
Comment (Jorropo) [reply p2p/net/swarm/swarm_stream.go:85]: Thx for explaining why it's private.
Just so you know this doesn't actually prevent anyone from implementing it.
You can use a badge using an internal package if you want to do.

File: p2p/net/swarm/swarm_stream_test.go
Comment (Jorropo) [reply p2p/net/swarm/swarm_stream_test.go]: I didn't thought of having more than one stream running at once but I think it's good idea.

My point was just about calling onDone out-of-order.

File: p2p/transport/webrtc/transport.go
Comment (Jorropo) [reply p2p/transport/webrtc/transport.go]: Dumb question but shouldn't udpmux also use 1460 MTU ?

Comment (sukunrt) [reply p2p/transport/webrtc/transport.go]: I think so. Will take that in a separate PR.

File: p2p/transport/webrtc/stream.go
Comment (Jorropo) [inline p2p/transport/webrtc/stream.go:150]: That weird.

**Review by Jorropo — DISMISSED** [review]

File: p2p/transport/webrtc/stream_write.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream_write.go:155]: The channel will keep the message, no? when the Write goroutine comes around to select block? 
There's a problem with concurrent Write calls but that's a race anyway

Comment (Jorropo) [reply p2p/transport/webrtc/stream_write.go:155]: I overlooked the fact that it is buffered.
It looks correct then, I would rather use already existing primitives than reimplementing what is well tested in the `sync` package but if you think this is fine then keep it.

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go:74]: Yeah, I don't remember why I did it this way.

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go:283]: Not sure if there's a race here but there's no point risking it.

File: p2p/transport/webrtc/stream_write.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream_write.go:155]: I don't mind changing it if the performance is better, but since this PR didn't introduce this particular structure, let's take this in a separate PR.

File: p2p/transport/webrtc/transport.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport.go]: This doesn't matter. It's just used for receiving and not for deciding the packet size on the write path. Renamed the variable and added a comment.

**Comment by p-shahi** [comment]
@Jorropo can you help review this PR?

File: p2p/transport/webrtc/stream.go
Comment (guillaumemichel) [inline p2p/transport/webrtc/stream.go:26]: where does this number come from? 

giving the rationale or link to reference would be great!

File: p2p/transport/webrtc/stream.go
Comment (guillaumemichel) [inline p2p/transport/webrtc/stream.go:47]: why 10 seconds?

**Review by guillaumemichel — APPROVED** [review]
LGTM!

It is my first time looking at the WebRTC implementation, so other reviews are also welcome.

**Comment by p-shahi** [comment]
go-libp2p open maintainers call: We have @guillaumemichel 's approval and pending other review comments we will merge this PR. There are other Pion fixes that are necessary to take WebRTC out of experimental as well but merging this will help get the next v0.33 release out the door

File: p2p/transport/webrtc/transport_test.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport_test.go:743]: I wanted to confirm whether quic and yamux also do this. They do. I have fixed it to return the correct error and removed the TODOs

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go:26]: This should be 50. Added an explanation. Please comment if it needs elaborating.

File: p2p/net/swarm/swarm_stream.go
Comment (Stebalien) [reply p2p/net/swarm/swarm_stream.go:85]: This seems totally unnecessary.  Why not just call `closeAndRemoveStream` immediately? From the perspective of libp2p, the stream is closed. If calling `closeAndRemoveStream` causes issues, that's a separate bug that needs to be fixed.


Really, _all_ stream closure is async, this isn't unique to webrtc.

**Review by Stebalien — CHANGES_REQUESTED** [review]

**Comment by sukunrt** [comment]
@Stebalien I have removed `AsyncClose`. 

The difference between webrtc and yamux/quic is that you're relying on the transport stack to discard all the buffers when you do a Close. This happens with yamux. With quic, the receive side of the stream is not immediately discarded but the peer doesn't get credits to open a new stream till it cleanly finishes the stream we have sent `STOP_SENDING` on. 

No such mechanism exists in webrtc. There is no limit to streams you can open(At least not in pion) or unilateral stream closing since streams are reusable. In webrtc `datachannel.Close` is a 1RTT sync procedure over the FIN / FIN_ACK closing that we do. So an effective `AsyncClose` procedure needs to do it after the datachannel is Closed and not after a FIN_ACK is received.  Moreover pion never drops a reference to the datachannel(including all the transport layer allocated buffers) till the connection is open. So a system like this only makes sense if we fix: https://github.com/pion/webrtc/issues/2672 

There is one way to handle this other than doing an explicit async close without waiting for 1RTT which would require us to check a misbehaving peer which is forcing us to keep all the datachannels open and just close the connection to such a peer. We can discuss that strategy in a separate issue. That'd also require https://github.com/pion/webrtc/issues/2672

**Comment by Stebalien** [comment]
> With quic, the receive side of the stream is not immediately discarded but the peer doesn't get credits to open a new stream till it cleanly finishes the stream after we have sent STOP_SENDING on.

Can't we do the same thing here? My objection is adding this hack to libp2p itself as this is a transport specific detail.

File: p2p/transport/webrtc/stream.go
Comment (guillaumemichel) [reply p2p/transport/webrtc/stream.go:26]: It is still `500` (not `50`) and I don't see the explanation

**Comment by sukunrt** [comment]
>My objection is adding this hack to libp2p itself as this is a transport specific detail.

That's a fair point. I'll open an issue to discuss.

**Review by guillaumemichel — APPROVED** [review]

**Comment by sukunrt** [comment]
@Stebalien 
>Why not just call closeAndRemoveStream immediately? 

I have addressed this and removed the `AsyncClose` logic. Is it fine to merge?

File: p2p/transport/webrtc/stream_read.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream_read.go:39]: nit: not related, but this locking is strange. We could write this as:

```go
s.mx.Unlock()
var msg pb.Message
err := s.reader.ReadMsg(&msg)
s.mx.Lock()
if err != nil {
    // ...
}
// no second "lock" call needed.

File: p2p/transport/webrtc/stream.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream.go:123]: nit: no need to initialize this.

File: p2p/transport/webrtc/stream.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream.go:143]: nit: same as the other case, I generally prefer:

```go
s.mx.Lock()
closed := s.closeForShutdownErr != nil
s.mx.Unlock()
...
```

File: p2p/transport/webrtc/stream.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream.go:191]: This seems like a strange way to do this. It'll also be racy (e.g., if the user calls the same function to change the read deadline).

Do we need to change the read deadline? If this is the only way to do this, let's add something to `SetReadDeadline` so it can't be changed again by the user.

File: p2p/transport/webrtc/stream.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream.go:225]: Future cleanup: make some form of `notifyStateChanged` function.

File: p2p/transport/webrtc/stream.go
Comment (Stebalien) [inline p2p/transport/webrtc/stream.go:263]: same race comment as above.

File: p2p/transport/webrtc/stream_write.go
Comment (Stebalien) [reply p2p/transport/webrtc/stream_write.go:156]: Using a channel as a "is this changed" flag is a common idiom, assuming there's only _one_ reader. `sync.Cond` is mostly useful when there are many readers.

File: p2p/transport/webrtc/transport_test.go
Comment (Stebalien) [inline p2p/transport/webrtc/transport_test.go:21]: Is this the rand you want to use?

**Review by Stebalien — COMMENTED** [review]
Some comments, mostly questions, otherwise LGTM.

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go:225]: I've replaced the three channels with 1 `writeStateChanged` and using a method to notify on the channel removing all the `select` blocks everywhere.

File: p2p/transport/webrtc/transport_test.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport_test.go]: I did mostly for this: https://github.com/golang/go/issues/21835

But looks like some of the changes have made it to math/rand and go1.23 will have a migrate tool to move to math/rand/v2. So the stdlib rand is fine here.

File: p2p/transport/webrtc/stream.go
Comment (sukunrt) [reply p2p/transport/webrtc/stream.go]: This is the only way to unblock a waiting reader. Even closing the datachannel won't unblock it till 1RTT later when the peer replies with a close datachannel on its part. 

I've changed this to make SetReadDeadline a noop after the read half is closed so the user can't change this again.

**Comment by Stebalien** [comment]
> I have addressed your concerns here. I want to do a release today. Happy to take any comments in a follow up PR and if something major is missed we can do a patch release.


Nope, LGTM!

File: p2p/transport/webrtc/transport.go
Comment (sukunrt) [reply p2p/transport/webrtc/transport.go]: I think I found the bug:
This was causing pion to send larger packets some times. 
https://github.com/pion/sctp/pull/301/files

**Pull request merged**
